### PR TITLE
Properly support CompletionStage as a return type in caching extension

### LINKED ISSUE #23816: IllegalStateException if method return type is CompletionStage<Xxx> using @CacheResult
### Describe the bug

Method return type is CompletionStage<Xxx> cause "The current thread cannot be blocked: vert.x-eventloop-thread-xx"

```java
  // raise exception
 @CacheResult(cacheName = "/api/m/test/stage")
 @GET
 @Path("/stage")
 @Produces(MediaType.TEXT_PLAIN)
 public CompletionStage<String> stage() {
      return CompletableFuture.completedFuture("stage:" + LocalDateTime.now());
 }
```

### Expected behavior
 
Should work when function return type CompletionStage<XXX> Annotated with  @CacheResult or  @CacheInvalidate

### Actual behavior

```java
13:12:33.165 DEBUG [vert.x-eventloop-thread-8] [null:-1,null()] RoutingContext failure (500): java.lang.IllegalStateException: The current thread cannot be blocked: vert.x-eventloop-thread-8
	at io.smallrye.mutiny.operators.uni.UniBlockingAwait.await(UniBlockingAwait.java:30)
	at io.smallrye.mutiny.groups.UniAwait.atMost(UniAwait.java:65)
	at io.smallrye.mutiny.groups.UniAwait.indefinitely(UniAwait.java:46)
	at io.quarkus.cache.runtime.CacheResultInterceptor.intercept(CacheResultInterceptor.java:115)
	at io.quarkus.cache.runtime.CacheResultInterceptor_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at com.xgosh.test.TestResource_Subclass.stage(Unknown Source)
	at com.xgosh.test.TestResource_ClientProxy.stage(Unknown Source)
	at com.xgosh.test.TestResource$quarkusrestinvoker$stage_971e8502f0474c9186922d25a243c5718389abdd.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
```

### How to Reproduce?

Test Code

```java
/**
 * <pre>
 * http://localhost:7070/api/m/test/normal
 * http://localhost:7070/api/m/test/uni
 * http://localhost:7070/api/m/test/stage
 * http://localhost:7070/api/m/test/stageBlocking
 * </pre>
 */
@RegisterForReflection
@ApplicationScoped
@Path("/api/m/test")
public class TestResource {

    /**
     * No error
     */
    @CacheResult(cacheName = "/api/m/test/normal")
    @GET
    @Path("/normal")
    @Produces(MediaType.TEXT_PLAIN)
    public String normal() {
        return "normal:" + LocalDateTime.now();
    }

    /**
     * No error
     */
    @CacheResult(cacheName = "/api/m/test/uni")
    @GET
    @Path("/uni")
    @Produces(MediaType.TEXT_PLAIN)
    public Uni<String> uni() {
        return Uni.createFrom().item("uni:" + LocalDateTime.now());
    }

    /**
     * Error
     */
    @CacheResult(cacheName = "/api/m/test/stage")
    @GET
    @Path("/stage")
    @Produces(MediaType.TEXT_PLAIN)
    public CompletionStage<String> stage() {
        return CompletableFuture.completedFuture("stage:" + LocalDateTime.now());
    }

    /**
     * No error
     */
    @CacheResult(cacheName = "/api/m/test/stageBlocking")
    @GET
    @Path("/stageBlocking")
    @Produces(MediaType.TEXT_PLAIN)
    @Blocking
    public CompletionStage<String> stageBlocking() {
        return CompletableFuture.completedFuture("stageBlocking:" + LocalDateTime.now());
    }

}
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

```txt
openjdk version "17.0.2" 2022-01-18
OpenJDK Runtime Environment GraalVM CE 21.3.1 (build 17.0.2+8-jvmci-21.3-b09)
OpenJDK 64-Bit Server VM GraalVM CE 21.3.1 (build 17.0.2+8-jvmci-21.3-b09, mixed mode, sharing)
```
### GraalVM version (if different from Java)

GraalVM CE 21.3.1 

### Quarkus version or git rev

2.7.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by geoand**
Fixes: #23816

### CODE REVIEW

File: extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java [hunk 1]
```diff
@@ -135,7 +137,39 @@ protected Object getCacheKey(Cache cache, List<Short> cacheKeyParameterPositions
...
+            return ReturnType.CompletionStage;
+        }
+        return ReturnType.NonAsync;
+    }
+
+    protected Uni<?> asyncInvocationResultToUni(Object invocationResult, ReturnType returnType) {
+        if (returnType == ReturnType.Uni) {
+            return (Uni<?>) invocationResult;
+        } else if (returnType == ReturnType.CompletionStage) {
+            return Uni.createFrom().completionStage((CompletionStage<?>) invocationResult);
```
Comment (gwenneg) [inline extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java:154]: I'm not sure `@CacheResult#lockTimeout` will work if the `Uni` is created from the `CompletionStage` here. Delaying the `CompletionStage` computation (with a `Supplier<CompletionStage>`) until the `Uni` is resolved may be necessary:
```suggestion
            return Uni.createFrom().completionStage(() -> (CompletionStage<?>) invocationContext.proceed());
```
(the real code would need a proper `Supplier` and not that lambda 😄)

**Review by gwenneg — COMMENTED** [review]
Thank you @geoand!

I'm not done reviewing the tests but first I have a few comments about the runtime code.

**Comment by geoand** [comment]
Thanks for the feedback.

I'll have a look tomorrow as I am taking the day off today

**Comment by geoand** [comment]
Comments addressed

**Comment by geoand** [comment]
This stuff will likely need to be rewritten for https://github.com/quarkusio/quarkus/pull/24894, but it's best to separate the two as this one addresses a clear bug, while the Kotlin corroutine support is a feature request

**Comment by gwenneg** [comment]
I should be able to finish this review later today, sorry for the delay 😃

**Comment by geoand** [comment]
Not a problem at all :)

File: extensions/cache/deployment/src/test/java/io/quarkus/cache/test/runtime/CompletionStageReturnTypeTest.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        Object value2 = cachedService.cacheResult2(KEY_2).get();
+
+        // We will invalidate some data (only KEY_1) in all caches later.
+        cachedService.cacheInvalidate(KEY_1).get();
+        // For now, the method that will invalidate the data should not be invoked, as CompletionStage is eager.
+        assertEquals(1, cachedService.getCacheInvalidateInvocations());
+
+        // The data for the second key should still be cached at this point.
+        Object value4 = cachedService.cacheResult2(KEY_2).get();
+        assertSame(value2, value4);
```
Comment (gwenneg) [inline extensions/cache/deployment/src/test/java/io/quarkus/cache/test/runtime/CompletionStageReturnTypeTest.java:90]:
```suggestion
        assertSame(value3, value4);
```

**Review by gwenneg — COMMENTED** [review]
Sorry for the delay 😃 

I noticed something wrong with `CacheResult#lockTimeout` and `Uni` while reviewing this PR, so I spent some time confirming that there is indeed a bug in that area. `CompletionStage` is affected in a similar way. I'll create an issue about it or a PR to fix it depending on how much time I have this week.

**Comment by gwenneg** [comment]
I hope I didn't break a test with the variables names changes 😄

**Comment by geoand** [comment]
Thanks a lot for the review, suggestions applied :)

**Review by gwenneg — APPROVED** [review]
LGTM

**Pull request merged**